File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/settings/edit_password/ChangePasswordFragment.java
package io.muun.apollo.presentation.ui.settings.edit_password;

import io.muun.apollo.R;
import io.muun.apollo.domain.errors.UserFacingError;
import io.muun.apollo.presentation.ui.activity.extension.MuunDialog;
import io.muun.apollo.presentation.ui.view.MuunButton;
import io.muun.apollo.presentation.ui.view.MuunTextInput;

import android.app.Activity;
import android.view.View;
import android.widget.CheckBox;
import butterknife.BindView;
import butterknife.OnCheckedChanged;
import butterknife.OnClick;

public class ChangePasswordFragment extends BaseEditPasswordFragment<ChangePasswordPresenter>
        implements ChangePasswordView {

    @BindView(R.id.change_password)
    MuunTextInput password;

    @BindView(R.id.change_password_condition)
    CheckBox condition;

    @BindView(R.id.change_password_continue)
    MuunButton continueButton;

    @Override
    protected void inject() {
        getComponent().inject(this);
    }

    @Override
    protected int getLayoutResource() {
        return R.layout.change_password_fragment;
    }

    @Override
    protected void initializeUi(View view) {
        super.initializeUi(view);

        password.setPasswordRevealEnabled(true);

        continueButton.setEnabled(false);
    }

    @Override
    public boolean onBackPressed() {

        final MuunDialog muunDialog = new MuunDialog.Builder()
                .title(R.string.change_password_abort_dialog_title)
                .message(R.string.change_password_abort_dialog_message)
                .positiveButton(R.string.change_password_abort_dialog_yes, this::abort)
                .negativeButton(R.string.change_password_abort_dialog_no, null)
                .build();

        getParentActivity().showDialog(muunDialog);

        return true;
    }

    private void abort() {
        safeGetParentActivity().ifPresent(Activity::finish);
    }

    @Override
    public void setLoading(boolean loading) {
        continueButton.setLoading(loading);
        condition.setEnabled(!loading);
    }

    @Override
    public void setPasswordError(UserFacingError error) {
        password.clearError();

        if (error != null) {
            password.setError(error);
            focusInput(password);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        focusInput(password);
    }

    @Override
    protected boolean blockScreenshots() {
        return true;
    }

    @OnCheckedChanged(R.id.change_password_condition)
    public void onConditionCheckedChanged() {
        continueButton.setEnabled(condition.isChecked());
    }

    @OnClick(R.id.change_password_continue)
    void onContinueButtonClick() {
        presenter.submitPassword(password.getText().toString());
    }
}

File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/settings/edit_password/ChangePasswordFragment.kt
package io.muun.apollo.presentation.ui.settings.edit_password

import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import androidx.viewbinding.ViewBinding
import io.muun.apollo.R
import io.muun.apollo.databinding.FragmentChangePasswordBinding
import io.muun.apollo.domain.errors.UserFacingError
import io.muun.apollo.presentation.ui.activity.extension.MuunDialog
import io.muun.apollo.presentation.ui.base.BaseActivity
import io.muun.apollo.presentation.ui.base.SingleFragment
import io.muun.apollo.presentation.ui.view.MuunButton
import io.muun.apollo.presentation.ui.view.MuunTextInput
import io.muun.common.Rules

class ChangePasswordFragment : SingleFragment<ChangePasswordPresenter>(), ChangePasswordView {

    private val binding: FragmentChangePasswordBinding
        get() = getBinding() as FragmentChangePasswordBinding

    private val passwordInput: MuunTextInput
        get() = binding.changePasswordInput

    private val passwordConfirmInput: MuunTextInput
        get() = binding.changePasswordConfirmInput

    private val condition: CheckBox
        get() = binding.changePasswordCondition

    private val continueButton: MuunButton
        get() = binding.changePasswordContinue

    override fun inject() {
        component.inject(this)
    }

    override fun getLayoutResource(): Int {
        return R.layout.fragment_change_password
    }

    override fun bindingInflater(): (LayoutInflater, ViewGroup, Boolean) -> ViewBinding {
        return FragmentChangePasswordBinding::inflate
    }

    override fun initializeUi(view: View) {
        continueButton.isEnabled = false

        passwordInput.setPasswordRevealEnabled(true)
        passwordInput.setOnChangeListener(this) {
            validateInputs()
        }
        passwordConfirmInput.setPasswordRevealEnabled(true)
        passwordConfirmInput.setOnChangeListener(this) {
            validateInputs()
        }

        condition.setOnCheckedChangeListener { _, _ ->
            validateInputs()
        }
        continueButton.setOnClickListener { onContinueButtonClick() }
    }

    override fun onBackPressed(): Boolean {
        val muunDialog = MuunDialog.Builder()
            .title(R.string.change_password_abort_dialog_title)
            .message(R.string.change_password_abort_dialog_message)
            .positiveButton(R.string.change_password_abort_dialog_yes) { abort() }
            .negativeButton(R.string.change_password_abort_dialog_no, null)
            .build()

        parentActivity.showDialog(muunDialog)
        return true
    }

    private fun abort() {
        safeGetParentActivity().ifPresent(BaseActivity<*>::finishActivity)
    }

    override fun setLoading(loading: Boolean) {
        continueButton.setLoading(loading)
        condition.isEnabled = !loading
    }

    override fun setPasswordError(error: UserFacingError?) {
        passwordInput.clearError()

        if (error != null) {
            passwordInput.setError(error)
            passwordInput.requestFocusInput()
        }
    }

    override fun setConfirmPasswordError(error: UserFacingError?) {
        passwordConfirmInput.clearError()

        if (error != null) {
            passwordConfirmInput.setError(error)
            passwordConfirmInput.requestFocusInput()
        }
    }

    override fun onResume() {
        super.onResume()
        passwordInput.requestFocusInput()
    }

    override fun blockScreenshots(): Boolean {
        return true
    }

    private fun validateInputs() {
        val validPassword = isValidPassword(passwordInput.text.toString())
        val validPasswordConfirm = isValidPassword(passwordConfirmInput.text.toString())

        continueButton.isEnabled = validPassword && validPasswordConfirm && condition.isChecked
    }

    private fun onContinueButtonClick() {
        presenter.submitPassword(
            passwordInput.text.toString(),
            passwordConfirmInput.text.toString()
        )
    }

    private fun isValidPassword(password: String) =
        !TextUtils.isEmpty(password) && password.length >= Rules.PASSWORD_MIN_LENGTH

}